Create a random console_unique_id

B3n30 opened the pull request:
This will allow that each citra gets its own console_unique_id and gets rid of the hardcoded 0xDEADC0DE.

It will create a random console_unique_id when the config savegame is created.
To also have to ability to change the console_unique_id without formatting the complete config savegame a button was added to the system config dialog which allows to generate a new random console_unique_id.
<img width="487" alt="system config dialog with new refresh console id button" src="https://cloud.githubusercontent.com/assets/26032316/25100392/31cd65c0-23b0-11e7-8e67-5199efa5191d.png">
One minor draw back is that this won't allow to set the console_unique_id manually. But most users won't ever need to set it manually, and those i need to set it, should know how to edit the savegame. On the other side adding a textbox for that in the ui, seemed way to much for a feature used that rarely.

Subv commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +667,13 @@ SoundOutputMode GetSoundOutputMode() {
 }
 
+void GenerateConsoleUniqueId() {
+        const u64 console_id = dis(rng);
```
There are too many whitespaces here

RavenHome1 commented:
is there any advantage to this?

B3n30 commented:
In my opinion that should be included as soon as possible. Every user using Citra now with the same unique id. If at some stage a unique id is required every user is forced to generate it manually. This might cause a lot of support requests and issues. Therefore adding a unique id to each Citra, shouldn't be delayed, cause right now there is no use for it.

jroweboy commented on src/citra_qt/configuration/configure_system.ui:
```diff
@@ -251,5 +267,12 @@
  </widget>
  <resources/>
- <connections/>
+ <connections>
```
I don't remember anywhere else in the program that we set up connections for GUI elements in the ui file. If we want to move to that it might be worth it to make a new issue about it.

B3n30 replied:
Moved the connect form .ui to the constructor, which is probably the better way. As i recently discussed with @JayFoxRox would it be best, if we get rid of the .ui files and implement everything in the cpp-files.

MerryMage commented on src/citra_qt/configuration/configure_system.ui:
```diff
@@ -220,6 +220,22 @@
+           <enum>Qt::RightToLeft</enum>
+          </property>
+          <property name="text">
+           <string>Refresh Console ID</string>
```
s/Refresh/Regenerate/ may be better wording.

JayFoxRox replied:
Agreed. Refresh is very confusing and users might accidentally hit it.
"Generate new console ID" would be better + there should probably be a warning (as this could potentially break games if they are all of sudden running on a different 3DS)

B3n30 replied:
These settings can't be changed while a game is running. So hitting this button will behave as if you plug in your game into another 3ds. I don't think a warning is needed

yuriks replied:
I agree that "Regenerate" is also a better word to use. "Refresh" sounds like something that should yield the same result if I click the button a bunch of times.

B3n30 replied:
It's now called "Generate new console ID"

MerryMage commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +665,11 @@ SoundOutputMode GetSoundOutputMode() {
+void GenerateConsoleUniqueId() {
+    const u64 console_id = dis(rng);
+    SetConfigInfoBlock(ConsoleUniqueID1BlockID, sizeof(console_id), 0xE, &console_id);
+    SetConfigInfoBlock(ConsoleUniqueID2BlockID, sizeof(console_id), 0xE, &console_id);
```
Would it be possible to return an error if these calls fail?

If this did, one could replace the code in `FormatConfig()` with a call to this function to avoid code duplication.

jroweboy commented on src/citra_qt/configuration/configure_system.cpp:
```diff
@@ -17,6 +17,7 @@ ConfigureSystem::ConfigureSystem(QWidget* parent) : QWidget(parent), ui(new Ui::
     ui->setupUi(this);
     connect(ui->combo_birthmonth, SIGNAL(currentIndexChanged(int)),
             SLOT(updateBirthdayComboBox(int)));
+    connect(ui->button_refresh_console_id, SIGNAL(clicked()), this, SLOT(refreshConsoleID()));
```
Minor nitpick, but we are slowly moving over to qt5 signals and slots syntax (as the qt4 stuff is so out of date and we don't really support it anymore) http://doc.qt.io/qt-5/signalsandslots.html It'd be nice to change that in the entire function, but if you don't feel like it, it'd be a simple task for a beginner to pick up.

lioncash commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -118,6 +118,9 @@ static const std::vector<u8> cfg_system_savedata_id = {
 
 static u32 preferred_region_code = 0;
 
+static std::mt19937_64 rng{};
```
These can just be placed within the two scopes they're needed in, rather than being file-scope, or put into an internally linked function to generate the value.

yuriks replied:
- This needs to be seeded somehow, otherwise it will always generate the same value.
- Rather than using a prng, use something like [random_device](http://en.cppreference.com/w/cpp/numeric/random/random_device) or a random source from cryptocpp. (iirc implementations of random_device tend to be very spotty.)

B3n30 replied:
std::random_device shouldn't be used, at least not if it should work properly on different platforms. 
http://en.cppreference.com/w/cpp/numeric/random/random_device:
> std::random_device may be implemented in terms of an implementation-defined pseudo-random number engine if a non-deterministic source (e.g. a hardware device) is not available to the implementation. In this case each std::random_device object may generate the same number sequence.
Which in this case would mean, hitting the button the first time would always result in the same console_id.
Thus I used the rng from cryptopp.

B3n30 commented:
Addressed all of the issues.
@JayFoxRox requested a warning dialog and a feedback that it actually changed the console_id
This is how it looks now:
<img width="201" alt="bildschirmfoto 2017-04-18 um 11 27 02" src="https://cloud.githubusercontent.com/assets/26032316/25124681/c0ebb50a-242c-11e7-962a-8d79d15e25d4.png">
Clicking the button will lead to:
<img width="426" alt="bildschirmfoto 2017-04-18 um 11 25 29" src="https://cloud.githubusercontent.com/assets/26032316/25124697/cf14390e-242c-11e7-9248-cbecba3bad6f.png">
And only if you click ok, the console ID will be changed and this message will appear:
<img width="431" alt="bildschirmfoto 2017-04-18 um 11 25 44" src="https://cloud.githubusercontent.com/assets/26032316/25124727/e4097b58-242c-11e7-969b-b8676e3ba4e8.png">

wwylele commented:
I doubt the complete randomness of this block, for
 - The config savegame dump from my 3DS has random-looking value for the first four bytes of the console unique ID block , but the second four bytes are "00 00 0f 80", which looks not random at all.
 - As 3dbrew says (and confirmed on my 3DS): The first word is the same as LocalFriendCodeSeed, while the latter is a separate word. 

I will try to find out what the second word is.

lioncash commented on src/citra_qt/configuration/configure_system.cpp:
```diff
@@ -140,3 +144,21 @@ void ConfigureSystem::updateBirthdayComboBox(int birthmonth_index) {
+                                  tr("This will replace your virtual 3DS with a new one. Your "
+                                     "current 3DS will not be recoverable. "
+                                     "This might have unexpected effects in games. Continue?"),
+                                  QMessageBox::Abort | QMessageBox::Yes);
```
This should likely be `QMessageBox::No` (reads nicer with the question in the message box instance) or `QMessageBox::Cancel`. Abort implies a currently started operation (which isn't the case here).

JayFoxRox commented on src/citra_qt/configuration/configure_system.cpp:
```diff
@@ -147,11 +147,11 @@ void ConfigureSystem::updateBirthdayComboBox(int birthmonth_index) {
-        QMessageBox::Abort | QMessageBox::Yes);
+    reply = QMessageBox::critical(this, tr("Warning"),
+                                  tr("This will replace your virtual 3DS with a new one. Your "
+                                     "current 3DS will not be recoverable. "
```
"current virtual 3DS"

(Otherwise we'll get support requests like: "will this break my 3DS?!?!" from idiots)

wwylele commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +658,30 @@ SoundOutputMode GetSoundOutputMode() {
 
+ResultCode GenerateConsoleUniqueId() {
+    CryptoPP::AutoSeededRandomPool rng;
+    const u32 random_number = static_cast<u16>((CryptoPP::Integer(rng, 16).ConvertToLong()));
```
casting to u16 and assign to u32?

B3n30 replied:
This was just to assure that only the lower 16bit are actually random while the upper 16bit are fixed to zero. This is the format this number is stored in block 0x90002. But the cast is probably unnecessary.

yuriks replied:
Masking with 0xFFFF would be clearer. In this case, the returned value will only have a range of [0, 2^16-1] anyway because you're passing 16 on the second param.

However, rather than using `Integer`, you should just call `rng.GenerateWord32(0, 0xFFFF)`.

yuriks replied:
This should also be a `u32_le` since you're writing it out as bytes to the block.

wwylele commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +658,30 @@ SoundOutputMode GetSoundOutputMode() {
+    const u64 console_id =
+        ((0x3FFFFFFFF) & local_friend_code_seed) | (static_cast<u64>(random_number) << 48);
+    ResultCode res =
+        SetConfigInfoBlock(ConsoleUniqueID1BlockID, sizeof(console_id), 0xE, &console_id);
```
The block need to be created by `CreateConfigInfoBlk` before set

B3n30 replied:
This is caused by an idea @MerryMage suggested. The format should have the create. The next commit will have a fix for it. There is still the problem that the system only creates new entries if the config savegame is written from scratch. Thus adding any block to the config savegame will require every user to delete the existing savegame. So i could add a create if the set fails, but a better solution would probably that a set should atomatically create it if it doesn't exist. This shouldn't be a separate PR.

wwylele commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +658,30 @@ SoundOutputMode GetSoundOutputMode() {
+ResultCode GenerateConsoleUniqueId() {
+    CryptoPP::AutoSeededRandomPool rng;
+    const u32 random_number = static_cast<u16>((CryptoPP::Integer(rng, 16).ConvertToLong()));
+    const u32 local_friend_code_seed = rng.GenerateWord32();
```
Shouldn't this be u64? Although yeah from my config savegame, the higher word seems not random but zero. However in the case that it is a u32, the next line `0x3FFFFFFFF &` (note there are eight F there) makes no sense be cause it won't mask anything out.

B3n30 replied:
According to https://3dbrew.org/wiki/Nand/private/movable.sed the upper 4 byte are usually zero. Also If u8[1] is zero, then u8 [0], [2], and [3] must be zero as well, with u8 being the upper 4byte. This can't be assured if they are randomly generated.

wwylele replied:
@B3n30 First of all, the sentence "LocalFriendCodeSeed, this is the same as the first 0x8-bytes from the movable.sed keyY. " in [this page](https://3dbrew.org/wiki/Nandrw/sys/LocalFriendCodeSeed_B) doesn't refer to offset 0 of movable.sed. It is more likely it refers to offset 0x118 instead (although this doesn't match the data I got from my 3ds at all). Second, let's assume this is a u32, then `0x3FFFFFFFF &` in the next line should be removed (which is there assuming a u64 seed)

IMO at this stage we should pause our work on citra and clean up the documentation on 3dbrew first. It looks to me like a mess and potentially containing mistakes.

B3n30 commented:
@lioncash I think that adding a button to the ui, is (almost) to much for a feature most users won't ever use. Adding an additional text seems - at least in my opinion - unnecessary, since the value itself is of no use for the user. The only important thing is, that its different, to the one it was before and that is different to any other console id used.

jfmherokiller commented:
the only issue i see with this commit is that it could make saves break. but i do recognize its purpose for increasing accuracy. I do have a suggestion though and it's to add an option to allow the user to import the console_id from a nand dump

B3n30 commented:
There is no save I know off, that will be get lost through this PR. This would only be saves tied to one specific console. If a user has dumped his config savegame from a 3DS, the best way (and also the easiest) is to just use that file as the config savegame. That way the citra console id will be the same as the real 3DS one. This behavior is the only reason the button exists: To get a unique id while you can still use your dumped savegame. Every user that didn't dump the savegame will automatically get a unique id, when the config savegame is created.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +658,30 @@ SoundOutputMode GetSoundOutputMode() {
+    CryptoPP::AutoSeededRandomPool rng;
+    const u32 random_number = static_cast<u16>((CryptoPP::Integer(rng, 16).ConvertToLong()));
+    const u32 local_friend_code_seed = rng.GenerateWord32();
+    const u64 console_id =
```
This should be a `u64_le` to be endian-safe.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +658,30 @@ SoundOutputMode GetSoundOutputMode() {
+}
+
+u64 GetConsoleUniqueId() {
+    u64 console_id;
```
Should also be a `u64_le`.

yuriks commented on src/citra_qt/configuration/configure_system.cpp:
```diff
@@ -140,3 +144,21 @@ void ConfigureSystem::updateBirthdayComboBox(int birthmonth_index) {
+
+void ConfigureSystem::refreshConsoleID() {
+    QMessageBox::StandardButton reply;
+    reply = QMessageBox::critical(this, tr("Warning"),
```
I assume you're splitting this in 2 lines because we align parameters to the opening parenthesis and it looks ugly like that? *barf* :(

yuriks commented on src/citra_qt/configuration/configure_system.cpp:
```diff
@@ -71,6 +75,10 @@ void ConfigureSystem::ReadSystemSettings() {
     ui->combo_sound->setCurrentIndex(sound_index);
+
+    // set the console id
+    u64_le console_id = Service::CFG::GetConsoleUniqueId();
```
Doesn't need to be `u64_le`. The *only* variables that need to be `_le` are one you're memcpying to or from, or otherwise interpreting as raw bytes.

yuriks commented on src/citra_qt/configuration/configure_system.cpp:
```diff
@@ -140,3 +148,21 @@ void ConfigureSystem::updateBirthdayComboBox(int birthmonth_index) {
+    if (reply == QMessageBox::No)
+        return;
+    u32_le random_number;
+    u64_le console_id;
```
These don't need to be `_le`

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +669,38 @@ SoundOutputMode GetSoundOutputMode() {
     return static_cast<SoundOutputMode>(block);
 }
 
+void GenerateConsoleUniqueId(u32_le& random_number, u64_le& console_id) {
```
Make the param types regular `u32` and `u64`, and use intermediary local `_le` variables to copy to.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +669,38 @@ SoundOutputMode GetSoundOutputMode() {
+        (local_friend_code_seed & 0x3FFFFFFFF) | (static_cast<u64_le>(random_number) << 48);
+}
+
+ResultCode SetConsoleUniqueId(u32_le random_number, u64_le console_id) {
```
Same, params should be non-le and copied to le local vars.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +669,38 @@ SoundOutputMode GetSoundOutputMode() {
+}
+
+u64_le GetConsoleUniqueId() {
+    u64 console_id;
```
This variable should be `_le`.

B3n30 commented:
fixed the `_le` usage. BTW, i would love to learn more about when to us _le and when to avoid it.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -669,37 +669,43 @@ SoundOutputMode GetSoundOutputMode() {
+void GenerateConsoleUniqueId(u32& random_number, u64& console_id) {
     CryptoPP::AutoSeededRandomPool rng;
-    random_number = rng.GenerateWord32(0, 0xFFFF);
+    u32_le random_number_le = rng.GenerateWord32(0, 0xFFFF);
```
This doesn't need to be `_le` since you don't copy it as bytes anywhere.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -669,37 +669,43 @@ SoundOutputMode GetSoundOutputMode() {
                       sizeof(local_friend_code_seed));
-    console_id =
-        (local_friend_code_seed & 0x3FFFFFFFF) | (static_cast<u64_le>(random_number) << 48);
+    u64_le console_id_le =
```
Doesn't need to be `_le`.

yuriks commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -669,37 +669,43 @@ SoundOutputMode GetSoundOutputMode() {
 
-ResultCode SetConsoleUniqueId(u32_le random_number, u64_le console_id) {
+ResultCode SetConsoleUniqueId(u32 random_number, u64 console_id) {
+    u32_le random_number_le = random_number;
```
Move this var down closer to where it's used.

wwylele commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -671,19 +671,15 @@ SoundOutputMode GetSoundOutputMode() {
-    random_number = random_number_le;
-    console_id = console_id_le;
+    console_id =
+        (local_friend_code_seed & 0x3FFFFFFFF) | (static_cast<u64_le>(random_number) << 48);
```
`static_cast<u64>`, so that endianess casting only happens at the assignment.

wwylele commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -663,5 +669,41 @@ SoundOutputMode GetSoundOutputMode() {
+    rng.GenerateBlock(reinterpret_cast<byte*>(&local_friend_code_seed),
+                      sizeof(local_friend_code_seed));
+    console_id =
+        (local_friend_code_seed & 0x3FFFFFFFF) | (static_cast<u64_le>(random_number) << 48);
```
static_cast<u64>

wwylele commented on src/core/hle/service/cfg/cfg.cpp:
```diff
@@ -438,13 +439,18 @@ ResultCode FormatConfig() {
-                              &CONSOLE_UNIQUE_ID);
+    u32_le random_number;
+    u64_le console_id;
+    GenerateConsoleUniqueId(random_number, console_id);
```
```
u32 random_number;
 u64 console_id;
GenerateConsoleUniqueId(random_number, console_id);
u32_le random_number_le = random_number; // use these two le variables to initialize config block
u64_le  console_id_le =  console_id;
```
Another solution is to let `GenerateConsoleUniqueId` returns a `tuple` and use `std::tie` for auto type conversion

bunnei commented:
Looks good. Thanks for submitting this!
